app/ota_client.py

```diff
@@ -896,48 +887,149 @@ def _gen_regular_files(self, rootfs_dir, regulars_file, target_dir):
+                elif vname[1] == "list":
+                    setattr(self, vname[-1], list(v))
+
+    def _process_regular_files(self, rootfs_dir, rfiles_list, target_dir):
+        with Manager() as manager:
+            ecb_queue = manager.Queue()
+            # variables passed to child processes
+            # variable naming pattern: <prefix>-<type>-<var_name>
+            # prefix tmp: for temporary use
+            # prefix staging: used to update corresponding class attribute
+            gvar_dict = {
+                "tmp-dict-hardlink_reg": manager.dict(),
+                "staging-dict-_rollback_dict": manager.dict(),
+            }
+
+            # default to one worker per CPU core
+            with Pool(
+                initializer=self._process_regular_files_pool_init, initargs=(gvar_dict,)
+            ) as pool:
+
+                # error_callback for workers
+                #   terminate all sub-processes
+                #   and pass exception to main process
+                def ecb(e):
+                    ecb_queue.put(e)
+                    pool.terminate()
+
+                for rfile_inf in rfiles_list:
+                    if (
+                        int(rfile_inf.links) >= 2
+                        and rfile_inf.sha256hash
+                        not in gvar_dict["tmp-dict-hardlink_reg"]
+                    ):
+                        # block the flow until the first copy of hardlinked file is ready
+                        try:
+                            pool.apply(
+                                self._process_regular_file,
+                                (rootfs_dir, target_dir, rfile_inf),
                             )
-                            if regular_inf.path.find("/boot/") == 0:
-                                # /boot directory file
-                                logger.debug(f"boot file: {regular_inf.path}")
-                                self._gen_boot_dir_file(
-                                    rootfs_dir, target_dir, regular_inf, prev_inf, fl
-                                )
-                            else:
-                                # others
-                                logger.debug(f"no boot file: {regular_inf.path}")
-                                self._gen_regular_file(
-                                    rootfs_dir, target_dir, regular_inf, prev_inf, fl
-                                )
-                            prev_inf = regular_inf
-                    except Exception as e:
-                        logger.exception("regular files setup error!:")
-                        res = False
-                    finally:
-                        os.chdir(cwd)
-        shutil.copy(log_file, "./regulars_dl.log")
-        return res
+                        except Exception as e:
+                            ecb(e)
+                            break
+                    else:
+                        try:
+                            pool.apply_async(
```

Comment:
@Bodong-Yang 
One question about apply_async.
apply_async returns AsyncResult object and my understanding is AsyncResult should be handled by caller to wait child process process.
But it seems this code doesn't wait child processes.
My understanding is correct?

Reply:
In line 994, **pool.join()** will block until all the tasks finished.

The worker process will start immediately after we issue pool.async_apply, and since the worker returns nothing, we don't need to deal with **AsyncResult**.

Reply:
I see, join waits the all processes. Understood.